### PR TITLE
Refactor Fuzzing and Increase Test Coverage

### PR DESCRIPTION
**Pull request by mhamilton723**

### CODE REVIEW

File: src/compute-model-statistics/src/test/scala/VerifyComputeModelStatistics.scala [hunk 1]
```diff
@@ -240,4 +241,32 @@ class VerifyComputeModelStatistics extends TestBase {
     assert(auc === cmsAUC)
   }
 
+  override def testObjects(): Seq[TestObject[ComputeModelStatistics]] = Seq(new TestObject(
+    new ComputeModelStatistics(),
+    {
+      val labelColumn = "label"
+      val predictionColumn = SchemaConstants.SparkPredictionColumn
+      val labelsAndPrediction = session.createDataFrame(
```
Comment (imatiach-msft) [inline src/compute-model-statistics/src/test/scala/VerifyComputeModelStatistics.scala]: maybe move the multiclass metrics dataset out and reuse it here and in the multiclass metrics test (this dataframe appears twice in the test file it looks like)

File: src/core/test/fuzzing/src/test/scala/Fuzzing.scala [hunk 2]
```diff
@@ -26,10 +29,53 @@ case class TestObject[S <: PipelineStage](stage: S,
...
+    def areEqual(a: DenseVector, b: Any): Boolean = b match {
+      case bArr:DenseVector =>
+        a.values.zip(bArr.values).forall {case (x, y) => doubleEq.areEqual(x, y)}
+    }
+  }
+
+  implicit lazy val rowEq: Equality[Row] = new Equality[Row]{
+    def areEqual(a: Row, bAny: Any): Boolean = bAny match {
+      case b:Row =>
+        if(a.length != b.length){return false}
```
Comment (imatiach-msft) [inline src/core/test/fuzzing/src/test/scala/Fuzzing.scala]: spacing is a little weird here (spaces before and after braces, parens)

File: src/core/test/fuzzing/src/test/scala/Fuzzing.scala [hunk 3]
```diff
@@ -26,10 +29,53 @@ case class TestObject[S <: PipelineStage](stage: S,
...
+            case lhs: Double =>
+              b(j).asInstanceOf[Double] === lhs
+            case lhs =>
+              lhs === b(j)
+          }
+        })
+    }
+  }
+
+  implicit lazy val dfEq: Equality[DataFrame] = new Equality[DataFrame]{
```
Comment (imatiach-msft) [inline src/core/test/fuzzing/src/test/scala/Fuzzing.scala]: nice!

File: src/core/test/fuzzing/src/test/scala/Fuzzing.scala [hunk 4]
```diff
@@ -26,10 +29,53 @@ case class TestObject[S <: PipelineStage](stage: S,
...
+        val b = ds.toDF()
+        if(a.columns !== b.columns){
+          return false
+        }
+        val aSort = a.collect()
+        val bSort = b.collect()
+        if (aSort.length != bSort.length){
+          return false
+        }
+        aSort.zip(bSort).forall {case (rowA, rowB) =>
```
Comment (imatiach-msft) [inline src/core/test/fuzzing/src/test/scala/Fuzzing.scala]: I don't think you can always do this -- the rows don't have any order in a dataframe.  So unless you do some expensive sort you can't really compare them like this.

File: src/core/test/fuzzing/src/test/scala/Fuzzing.scala [hunk 5]
```diff
@@ -123,40 +175,48 @@ trait SerializationFuzzing[S <: PipelineStage with MLWritable] extends FuzzingMe
...
+          assert(e1.fit(fitDF).transform(transDF) === e2.fit(fitDF).transform(transDF))
         case (t1: Transformer, t2: Transformer) =>
-          assert(compareDFs(t1.transform(transDF), t2.transform(transDF)))
+          assert(t1.transform(transDF) === t2.transform(transDF))
         case _ => throw new IllegalArgumentException(s"$stage and $loadedStage do not have proper types")
       }
       ()
     } finally {
-      FileUtils.forceDelete(new File(path))
+      try{
```
Comment (imatiach-msft) [inline src/core/test/fuzzing/src/test/scala/Fuzzing.scala]: (minor: spacing for braces)

File: src/core/test/fuzzing/src/test/scala/Fuzzing.scala [hunk 6]
```diff
@@ -123,40 +175,48 @@ trait SerializationFuzzing[S <: PipelineStage with MLWritable] extends FuzzingMe
...
+        FileUtils.forceDelete(new File(path))
+      }catch{
+        case _: FileNotFoundException =>
```
Comment (imatiach-msft) [inline src/core/test/fuzzing/src/test/scala/Fuzzing.scala]: was this meant to be ignored?  Maybe we should at least print something in this case to warn the user that the delete failed?

File: src/core/test/fuzzing/src/test/scala/FuzzingTest.scala [hunk 7]
```diff
@@ -24,74 +24,34 @@ class FuzzingTest extends TestBase {
...
-      "com.microsoft.ml.spark.CNTKModel",
       "com.microsoft.ml.spark.ValueIndexerModel",
-      "com.microsoft.ml.spark.PartitionSample",
-      "com.microsoft.ml.spark.Cacher",
-      "com.microsoft.ml.spark.ImageSetAugmenter",
-      "com.microsoft.ml.spark.EnsembleByKey",
-      "com.microsoft.ml.spark.ClassBalancer",
-      "com.microsoft.ml.spark.Timer")
+      "com.microsoft.ml.spark.CNTKLearner",
+      "com.microsoft.ml.spark.TrainClassifier",
```
Comment (imatiach-msft) [inline src/core/test/fuzzing/src/test/scala/FuzzingTest.scala]: what were the issues for the ones that are left?

File: src/core/utils/src/main/scala/JarLoadingUtils.scala [hunk 8]
```diff
@@ -50,9 +50,23 @@ object JarLoadingUtils {
 
   private val jarURLs = jarFileLocs.map(_.toURI.toURL)
 
-  val classLoader = new URLClassLoader(jarURLs.union(testOutputDirs
-    .map(file => new File(file.getCanonicalPath).toURI.toURL)).toArray,
-    this.getClass.getClassLoader)
+  private def addURL(url: URL): Unit = {
```
Comment (imatiach-msft) [inline src/core/utils/src/main/scala/JarLoadingUtils.scala]: nice fix!

File: src/partition-sample/src/test/scala/VerifyPartitionSample.scala [hunk 9]
```diff
@@ -64,4 +62,9 @@ class PartitionSampleSmokeTests extends TestBase {
       .toDF("Col1", "Col2")
   }
 
+  override def testObjects(): Seq[TestObject[PartitionSample]] = Seq(new TestObject[PartitionSample](
+    new PartitionSample().setMode("Head").setCount(3),
+    makeDF))
```
Comment (imatiach-msft) [inline src/partition-sample/src/test/scala/VerifyPartitionSample.scala]: minor style: you can put the makeDF on line above

File: src/summarize-data/src/test/scala/SummarizeDataSuite.scala [hunk 10]
```diff
@@ -29,6 +27,7 @@ class SummarizeDataSuite extends TestBase {
     val input = makeBasicDF()
     val summary = new SummarizeData()
     val result = summary.transform(input)
+    result.show()
```
Comment (imatiach-msft) [inline src/summarize-data/src/test/scala/SummarizeDataSuite.scala]: remove the show() in the test case

File: src/value-indexer/src/main/scala/ValueIndexer.scala [hunk 11]
```diff
@@ -151,7 +149,10 @@ class ValueIndexerModel(val uid: String)
   private def addCategoricalColumn[T: TypeTag](dataset: Dataset[_])
                                               (implicit ct: ClassTag[T]): DataFrame = {
     val nonNullLevels = getLevels.filter(_ != null)
-    val castLevels = nonNullLevels.map(_.asInstanceOf[T])
+    val castLevels = nonNullLevels.map {
+      case l: scala.math.BigInt => l.toInt.asInstanceOf[T]
```
Comment (imatiach-msft) [inline src/value-indexer/src/main/scala/ValueIndexer.scala]: why is the BigInt case needed in the map?

**Comment by imatiach-msft** [comment]
Looks good, just a couple minor comments

File: src/core/test/fuzzing/src/test/scala/Fuzzing.scala (hunk 6 above)
Comment (mhamilton723) [reply src/core/test/fuzzing/src/test/scala/Fuzzing.scala]: This effectively tries to delete a file if it exists and if not it doesent worry about it. making it into a try catch allows the previous error message to be surfaced

File: src/core/test/fuzzing/src/test/scala/FuzzingTest.scala (hunk 7 above)
Comment (mhamilton723) [reply src/core/test/fuzzing/src/test/scala/FuzzingTest.scala]: My laziness lol

File: src/value-indexer/src/main/scala/ValueIndexer.scala (hunk 11 above)
Comment (mhamilton723) [reply src/value-indexer/src/main/scala/ValueIndexer.scala]: BigInt does not cast to an Int, it needs to be converted before coercion.

File: src/core/test/fuzzing/src/test/scala/Fuzzing.scala (hunk 2 above)
Comment (mhamilton723) [reply src/core/test/fuzzing/src/test/scala/Fuzzing.scala]: Done

File: src/core/test/fuzzing/src/test/scala/Fuzzing.scala (hunk 4 above)
Comment (mhamilton723) [reply src/core/test/fuzzing/src/test/scala/Fuzzing.scala]: Fixed with a sort

File: src/core/test/fuzzing/src/test/scala/Fuzzing.scala (hunk 5 above)
Comment (mhamilton723) [reply src/core/test/fuzzing/src/test/scala/Fuzzing.scala]: Done

File: src/partition-sample/src/test/scala/VerifyPartitionSample.scala (hunk 9 above)
Comment (mhamilton723) [reply src/partition-sample/src/test/scala/VerifyPartitionSample.scala]: Done

File: src/summarize-data/src/test/scala/SummarizeDataSuite.scala (hunk 10 above)
Comment (mhamilton723) [reply src/summarize-data/src/test/scala/SummarizeDataSuite.scala]: DOne

File: src/compute-model-statistics/src/test/scala/VerifyComputeModelStatistics.scala (hunk 1 above)
Comment (mhamilton723) [reply src/compute-model-statistics/src/test/scala/VerifyComputeModelStatistics.scala]: Done

File: src/value-indexer/src/main/scala/ValueIndexer.scala (hunk 11 above)
Comment (imatiach-msft) [reply src/value-indexer/src/main/scala/ValueIndexer.scala]: I don't think there would ever be a BigInt in this case though.  Was there an exception that was thrown that led you to add this case?  Can the case be removed?

File: src/compute-model-statistics/src/test/scala/VerifyComputeModelStatistics.scala (hunk 1 above)
Comment (imatiach-msft) [reply src/compute-model-statistics/src/test/scala/VerifyComputeModelStatistics.scala]: it looks like it is still not reused (?)

Comment (imatiach-msft) [reply src/compute-model-statistics/src/test/scala/VerifyComputeModelStatistics.scala]: ah, you haven't pushed a new commit yet, will wait for it...

File: src/value-indexer/src/main/scala/ValueIndexer.scala (hunk 11 above)
Comment (mhamilton723) [reply src/value-indexer/src/main/scala/ValueIndexer.scala]: yes I had to add this to fix an error

File: src/core/test/fuzzing/src/test/scala/Fuzzing.scala [hunk 12]
```diff
@@ -123,40 +175,48 @@ trait SerializationFuzzing[S <: PipelineStage with MLWritable] extends FuzzingMe
       val loadedStage = reader.load(path)
```
Comment (imatiach-msft) [inline src/core/test/fuzzing/src/test/scala/Fuzzing.scala]: maybe add assert after the save since we ignore the file not found exception now

Comment (mhamilton723) [reply src/core/test/fuzzing/src/test/scala/Fuzzing.scala]: Done

**Pull request merged**